sampler_remote_test.go

```diff
@@ -17,18 +17,114 @@ package jaeger
 import (
 	"errors"
 	"fmt"
+	"sync"
 	"testing"
 	"time"
 
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
+	"github.com/uber/jaeger-lib/metrics"
 	mTestutils "github.com/uber/jaeger-lib/metrics/metricstest"
 
 	"github.com/uber/jaeger-client-go/log"
 	"github.com/uber/jaeger-client-go/testutils"
 	"github.com/uber/jaeger-client-go/thrift-gen/sampling"
 )
 
+func TestRemotelyControlledSampler_updateRace(t *testing.T) {
+	m := &Metrics{
+		SamplerRetrieved: metrics.NullCounter,
+		SamplerUpdated:   metrics.NullCounter,
+	}
+	initSampler, _ := NewProbabilisticSampler(0.123)
+	logger := log.NullLogger
+	fetcher := &testSamplingStrategyFetcher{response: []byte("probabilistic")}
+	parser := new(testSamplingStrategyParser)
+	updaters := []SamplerUpdater{new(ProbabilisticSamplerUpdater)}
+	sampler := NewRemotelyControlledSampler(
+		"test",
+		SamplerOptions.Metrics(m),
+		SamplerOptions.MaxOperations(42),
+		SamplerOptions.OperationNameLateBinding(true),
+		SamplerOptions.InitialSampler(initSampler),
+		SamplerOptions.Logger(logger),
+		SamplerOptions.SamplingServerURL("my url"),
+		SamplerOptions.SamplingRefreshInterval(time.Millisecond),
+		SamplerOptions.SamplingStrategyFetcher(fetcher),
+		SamplerOptions.SamplingStrategyParser(parser),
+		SamplerOptions.Updaters(updaters...),
+	)
+
+	s := makeSpan(1, "test")
+
+	var wg sync.WaitGroup
+
+	wg.Add(1)
```

Comment:
This test will only run each method one time.  Let's run them all repeatedly for a short period of time like the span test. I just ran this locally and it passes:

```
...
	end := make(chan struct{})
	accessor := func(f func()) {
		for {
			select {
			case <-end:
				return
			default:
				f()
			}
		}
	}

	go accessor(func() {
		sampler.UpdateSampler()
	})

	go accessor(func() {
		sampler.IsSampled(TraceID{Low: 1}, "test")
	})

	go accessor(func() {
		sampler.OnCreateSpan(s)
	})

	go accessor(func() {
		sampler.OnSetTag(s, "test", 1)
	})

	go accessor(func() {
		sampler.OnFinishSpan(s)
	})

	go accessor(func() {
		sampler.OnSetOperationName(s, "test")
	})

	time.Sleep(100 * time.Millisecond)
	close(end)
	sampler.Close()
```

Reply:
> This test will only run each method one time. Let's run them all repeatedly for a short period of time like the span test. I just ran this locally and it passes:
> 

Did you mean you ran it with commented fix? 
As far as I know race detector need memory accessed once from different goroutines to detect race condition.
When I comment fix:
```
func (s *RemotelyControlledSampler) OnCreateSpan(span *Span) SamplingDecision {
	//s.RLock()
	//defer s.RUnlock()
	return s.sampler.OnCreateSpan(span)
}
```
Test fails with
```
==================
WARNING: DATA RACE
Read at 0x00c00011a678 by goroutine 81:
  github.com/uber/jaeger-client-go.(*ProbabilisticSampler).IsSampled()
```

Reply:
You are right. With old version of code test does not fail. I've changed test so it looks like one in span_test.go.
Thank you!